File: app/src/main/java/com/stho/software/nyota/universe/AbstractSatellite.java
package com.stho.software.nyota.universe;

import com.stho.software.nyota.utilities.Angle;
import com.stho.software.nyota.utilities.Degree;
import com.stho.software.nyota.utilities.Formatter;
import com.stho.software.nyota.utilities.Hour;
import com.stho.software.nyota.utilities.IconNameValueList;
import com.stho.software.nyota.utilities.IconValueList;
import com.stho.software.nyota.utilities.Location;
import com.stho.software.nyota.utilities.Moment;
import com.stho.software.nyota.R;
import com.stho.software.nyota.utilities.Topocentric;
import com.stho.software.nyota.utilities.Vector;

/**
 * Created by shoedtke on 24.01.2017.
 */

public abstract class AbstractSatellite implements IElement {

    protected TLE tle;
    protected String name;
    protected Topocentric topocentric;
    protected Vector position = new Vector();
    protected Vector velocity = new Vector();
    protected Location location = new Location(0, 0, 0);

    public TLE getTLE() {
        return this.tle;
    }
    
    @Override
    public String getName() { return this.name; }

    abstract public int getImageId();

    @Override
    public int getLargeImageId() { return this.getImageId(); }

    abstract public void update(double julianDay);

    @Override
    public Topocentric getPosition() {
        return this.topocentric;
    }

    public Location getLocation() {
        return this.location;
    }

    /**
     * Speed in km/h
     * @return
     */
    public double getSpeed() { return 60 * this.velocity.getLength(); }


    public IconValueList getBasics(Moment moment) {
        IconValueList basics = new IconValueList();

        basics.add(R.drawable.horizontal, Angle.toString(topocentric.azimuth, Angle.AngleType.AZIMUTH) + Formatter.SPACE + Angle.toString(topocentric.altitude, Angle.AngleType.ALTITUDE));

        return basics;
    }

    public IconNameValueList getDetails(Moment moment) {
        IconNameValueList details = new IconNameValueList();

        details.add(R.drawable.horizontal, "Azimuth", Hour.fromDegree(topocentric.azimuth));
        details.add(R.drawable.horizontal, "Altitude", Degree.fromDegree(topocentric.altitude));
        details.add(R.drawable.equatorial, "Latitude", Degree.fromDegree(location.getLatitude()));
        details.add(R.drawable.equatorial, "Longitude", Degree.fromDegree(location.getLongitude()));
        details.add(R.drawable.equatorial, "Height", Formatter.df0.format(location.getAltitude()) + " km");
        details.add(R.drawable.distance, "Distance", Formatter.df0.format(topocentric.distance) + " km");
        details.add(R.drawable.star, "Speed", Formatter.df0.format(this.getSpeed()) + " km/h");
        details.add(R.drawable.time, "TLE", Formatter.date.format(this.tle.getDate()));

        return details;
    }


    public int getVisibility() {
        if (position != null && (topocentric.altitude > 5 && topocentric.altitude < 175))
            return R.drawable.visible;

        if (position != null && (topocentric.altitude < 0 || topocentric.altitude > 180))
            return R.drawable.invisible;

        return R.drawable.dizzy;
    }
}

File: app/src/main/java/com/stho/nyota/sky/universe/AbstractSatellite.kt
package com.stho.nyota.sky.universe

import com.stho.nyota.sky.utilities.*
import com.stho.nyota.sky.utilities.Formatter
import com.stho.nyota.sky.utilities.Vector

/**
 * Created by shoedtke on 24.01.2017.
 */
abstract class AbstractSatellite : IElement {

    abstract val tle: TLE

    override var position: Topocentric? = null
        protected set

    protected val positionVector: Vector = Vector()

    protected val velocity: Vector = Vector()

    var location: Location = Location(0.0, 0.0, 0.0)

    abstract fun updateFor(julianDay: Double)

    /**
     * Speed in km/h
     * @return
     */
    val speed: Double
        get() = 60 * velocity.length

    override fun getBasics(moment: Moment): PropertyList =
        PropertyList().apply {
            add(com.stho.nyota.R.drawable.compass, "Direction", position!!.toString())
            add(com.stho.nyota.R.drawable.horizontal, "Azimuth", Hour.fromDegree(position!!.azimuth))
            add(com.stho.nyota.R.drawable.horizontal, "Altitude", Degree.fromDegree(position!!.altitude))
            add(com.stho.nyota.R.drawable.equatorial, "Height", Formatter.toDistanceKmString(location.altitude))
            add(com.stho.nyota.R.drawable.distance, "Distance", Formatter.toDistanceKmString(position!!.distance))
            add(com.stho.nyota.R.drawable.empty, "Speed", Formatter.toSpeedString(speed))
        }

    override fun getDetails(moment: Moment): PropertyList =
        PropertyList().apply {
            add(com.stho.nyota.R.drawable.equatorial, "Latitude", Degree.fromDegree(location.latitude))
            add(com.stho.nyota.R.drawable.equatorial, "Longitude", Degree.fromDegree(location.longitude))
            add(com.stho.nyota.R.drawable.compass, "NORAD", tle.noradSatelliteNumber.toString())
            add(com.stho.nyota.R.drawable.empty, "TLE", Formatter.toString(tle.date, Formatter.timeZoneGMT, Formatter.TimeFormat.DATETIME_TIMEZONE))
            add(com.stho.nyota.R.drawable.empty, "Mean Distance", Formatter.toDistanceKmString(tle.meanDistanceFromEarth))
            add(com.stho.nyota.R.drawable.empty, "Revolutions", Formatter.df2.format(tle.revolutionsPerDay))
        }

    override val visibility: Int
        get() = when {
            Topocentric.isVisible(position) -> com.stho.nyota.R.drawable.visible
            Topocentric.isBelowHorizon(position) -> com.stho.nyota.R.drawable.invisible
            else -> com.stho.nyota.R.drawable.dizzy
        }

    override val isVisible: Boolean
        get() = Topocentric.isVisible(position)

    override fun isNear(otherPosition: Topocentric, toleranceInDegree: Double): Boolean =
        position?.isNear(otherPosition, toleranceInDegree) ?: false

    override fun isNear(otherPosition: Topocentric, azimuthTolerance: Double, altitudeTolerance: Double): Boolean =
        position?.isNear(otherPosition, azimuthTolerance, altitudeTolerance) ?: false

    override fun distanceTo(otherPosition: Topocentric): Double =
        position?.distanceTo(otherPosition) ?: Topocentric.INVALID_DISTANCE

}